### ISSUE TITLE
dokku restart consistently fails

### ISSUE THREAD

**Issue by matthewmueller** [dokku/dokku#3396]
## Description of problem

This might be similar to #3152, but I have it consistently failing. This may lead to a break-through. 

I was doing this to test restarts because I've had downtime with dokku in the past and I wasn't sure why given the restart policy configuration. I'm a total dokku fanboy, but this is a bit of a dealbreaker for me 😔. I've looked through the source quite a bit, but I haven't figured it out yet. I'm going to keep looking through.

### How reproducible

Two files:

**index.js**

```js
const express = require('express')
const app = express()

app.get('/', function(req, res) {
  res.send('hi world!')
})

app.get('/die', function(req, res) {
  process.exit(1)
})

app.listen(process.env.PORT, function() {
  console.log(`listening on ${process.env.PORT}`)
})
```

**package.json**

```json
{
  "name": "node-example",
  "engines": {
    "node": "10.x"
  },
  "scripts": {
    "start": "node index.js"
  },
  "dependencies": {
    "express": "4.16.4"
  }
}
```

### Steps to Reproduce

1. Launch this basic app on a dokku instance
2. Go to `/` just to make sure it's running
3. Go to `/die`, this will intentionally kill the process. This is to simulate a memory leak or 100 other reasons a node.js server can go down.
4. Look at `docker ps` on the server, you'll see it's re-run the container
5. Try going back to `/`, nginx status page 😭

#### Actual Results

We're sorry, but something went wrong.

#### Expected Results

hi world!

### `dokku report APP_NAME` output

```
-----> uname: Linux ip-172-31-14-74 4.15.0-1029-aws #30-Ubuntu SMP Wed Nov 21 23:58:13 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
-----> memory:
                     total        used        free      shared  buff/cache   available
       Mem:            983         208         113           0         661         619
       Swap:             0           0           0
-----> docker version:
       Client:
        Version:           18.09.1
        API version:       1.39
        Go version:        go1.10.6
        Git commit:        4c52b90
        Built:             Wed Jan  9 19:35:31 2019
        OS/Arch:           linux/amd64
        Experimental:      false

       Server: Docker Engine - Community
        Engine:
         Version:          18.09.1
         API version:      1.39 (minimum version 1.12)
         Go version:       go1.10.6
         Git commit:       4c52b90
         Built:            Wed Jan  9 19:02:44 2019
         OS/Arch:          linux/amd64
         Experimental:     false
-----> docker daemon info:
WARNING: No swap limit support
       Containers: 10
        Running: 1
        Paused: 0
        Stopped: 9
       Images: 9
       Server Version: 18.09.1
       Storage Driver: overlay2
        Backing Filesystem: extfs
        Supports d_type: true
        Native Overlay Diff: true
       Logging Driver: json-file
       Cgroup Driver: cgroupfs
       Plugins:
        Volume: local
        Network: bridge host macvlan null overlay
        Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
       Swarm: inactive
       Runtimes: runc
       Default Runtime: runc
       Init Binary: docker-init
       containerd version: 9754871865f7fe2f4e74d43e2fc7ccd237edcbce
       runc version: 96ec2177ae841256168fcf76954f7177af9446eb
       init version: fec3683
       Security Options:
        apparmor
        seccomp
         Profile: default
       Kernel Version: 4.15.0-1029-aws
       Operating System: Ubuntu 18.04.1 LTS
       OSType: linux
       Architecture: x86_64
       CPUs: 1
       Total Memory: 983.9MiB
       Name: ip-172-31-14-74
       ID: 5XRY:OHNV:IUOW:JBV5:MYSS:7CFN:H5VD:I7HR:TGFN:RJKA:J6BB:Q3C3
       Docker Root Dir: /var/lib/docker
       Debug Mode (client): true
       Debug Mode (server): false
       Registry: https://index.docker.io/v1/
       Labels:
       Experimental: false
       Insecure Registries:
        127.0.0.0/8
       Live Restore Enabled: false
       Product License: Community Engine

-----> sigil version: 0.4.0
-----> herokuish version:
       herokuish: 0.4.6
       buildpacks:
         heroku-buildpack-multi     v1.0.0
         heroku-buildpack-ruby      v197
         heroku-buildpack-nodejs    v134
         heroku-buildpack-clojure   v82
         heroku-buildpack-python    v146
         heroku-buildpack-java      v62
         heroku-buildpack-gradle    v27
         heroku-buildpack-scala     v83
         heroku-buildpack-play      v26
         heroku-buildpack-php       v148
         heroku-buildpack-go        v98
         buildpack-nginx            v10
-----> dokku version: 0.14.1
-----> dokku plugins:
       plugn: 0.3.0
         00_dokku-standard    0.14.1 enabled    dokku core standard plugin
         20_events            0.14.1 enabled    dokku core events logging plugin
         app-json             0.14.1 enabled    dokku core app-json plugin
         apps                 0.14.1 enabled    dokku core apps plugin
         build-env            0.14.1 enabled    dokku core build-env plugin
         certs                0.14.1 enabled    dokku core certificate management plugin
         checks               0.14.1 enabled    dokku core checks plugin
         common               0.14.1 enabled    dokku core common plugin
         config               0.14.1 enabled    dokku core config plugin
         docker-options       0.14.1 enabled    dokku core docker-options plugin
         domains              0.14.1 enabled    dokku core domains plugin
         enter                0.14.1 enabled    dokku core enter plugin
         git                  0.14.1 enabled    dokku core git plugin
         logs                 0.14.1 enabled    dokku core logs plugin
         network              0.14.1 enabled    dokku core network plugin
         nginx-vhosts         0.14.1 enabled    dokku core nginx-vhosts plugin
         plugin               0.14.1 enabled    dokku core plugin plugin
         proxy                0.14.1 enabled    dokku core proxy plugin
         ps                   0.14.1 enabled    dokku core ps plugin
         repo                 0.14.1 enabled    dokku core repo plugin
         scheduler-docker-local 0.14.1 enabled    dokku core scheduler-docker-local plugin
         shell                0.14.1 enabled    dokku core shell plugin
         ssh-keys             0.14.1 enabled    dokku core ssh-keys plugin
         storage              0.14.1 enabled    dokku core storage plugin
         tags                 0.14.1 enabled    dokku core tags plugin
         tar                  0.14.1 enabled    dokku core tar plugin
=====> node app information
       App dir:                       /home/dokku/node
       Git sha:                       9d61b85
       Deploy source:                 git
       Locked:                        false
=====> node ssl information
       Ssl dir:                       /home/dokku/node/tls
       Ssl enabled:                   false
       Ssl hostnames:
       Ssl expires at:
       Ssl issuer:
       Ssl starts at:
       Ssl subject:
       Ssl verified:
=====> node checks information
       Checks disabled list:          none
       Checks skipped list:           none
=====> node docker options information
       Docker options build:
       Docker options deploy:         --restart=on-failure:10
       Docker options run:
=====> node domains information
       Domains app enabled:           false
       Domains app vhosts:
       Domains global enabled:        false
       Domains global vhosts:
=====> node git information
       Git rev env var:               GIT_REV
       Git deploy branch:             master
       Git global deploy branch:      master
=====> node network information
       Network bind all interfaces:   false
       Network listeners:             172.17.0.3:5000
=====> node proxy information
       Proxy enabled:                 true
       Proxy type:                    nginx
       Proxy port map:                http:40944:5000
=====> node ps information
       Processes:                     1
       Deployed:                      true
       Running:                       true
       Restore:                       true
       Restart policy:                on-failure:10
       Status web.1:                  running    (CID: 1db3ff3b09a8)
=====> node scheduler-docker-local information
       Scheduler docker local disable chown:
=====> node storage information
       Storage build mounts:
       Storage deploy mounts:
       Storage run mounts:
```

### How (deb/make/rpm) and where (AWS, VirtualBox, physical, etc.) was Dokku installed?:

used this:

```
#!/bin/bash

wget https://raw.githubusercontent.com/dokku/dokku/v0.14.1/bootstrap.sh;
sudo DOKKU_TAG=v0.14.1 bash bootstrap.sh
```

on an amazon ec2 ubuntu box, ubuntu-bionic-18.04-amd64-server.

### Additional information

Result of docker inspect in the gist: https://gist.github.com/matthewmueller/c0a80bc2c883c1d3ac147e4dfe2abc57

---

Thanks in advance!

**Comment by josegonzalez** [comment]
Does running the following fix the issue?

```shell
dokku nginx:build-config APP
```

**Comment by matthewmueller** [comment]
Good call! that does seem to solve it. I guess it's the same issue as #3152. I'll close this one.

It doesn't make sense to me that docker would reassign the ports. I feel like that would be a much wider issue than dokku. I'll keep digging too!

**Comment by josegonzalez** [comment]
It reassigns IP addresses.

**Comment by josegonzalez** [comment]
Note: this was fixed in 0.20.2.
